### PR TITLE
SALTO-3986: omit circular transitions from deployment and warn

### PR DESCRIPTION
**Pull request by IdoZakk**
added a change validator to warn about circular transitions and a filter to prevent their deployment

---
_Release Notes_: 
* Jira Adapter:
Fixed a bug where circular workflow transitions prevented deployment

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/jira-adapter/src/filters/workflow/circular_transitions_filter.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      .filter(isAdditionOrModificationChange)
+      .map(getChangeData)
+      .filter(instance => instance.elemID.typeName === WORKFLOW_TYPE_NAME)
+      .filter(instance => _.isArray(instance.value.transitions))
+      .forEach(removeCircularTransitions)
+  },
+  // onDeploy: async changes => {
+  // },
+})
+export default filter
```
Comment (IdoZakk) [inline packages/jira-adapter/src/filters/workflow/circular_transitions_filter.ts:41]: I did not return the transitions after the deployment as they were deleted from the wrokflow

File: packages/jira-adapter/src/filters/workflow/circular_transitions_filter.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// This filter removes circular transitions that cannot be deployed
+const filter: FilterCreator = () => ({
+  name: 'circularTransitionsFilter',
+  preDeploy: async changes => {
+    changes
+      .filter(isInstanceChange)
+      .filter(isAdditionOrModificationChange)
+      .map(getChangeData)
+      .filter(instance => instance.elemID.typeName === WORKFLOW_TYPE_NAME)
+      .filter(instance => _.isArray(instance.value.transitions))
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/workflow/circular_transitions_filter.ts:35]: Same, you can use here `isWorkflowInstace`

**Review by alonstern — COMMENTED** [review]
👌

File: packages/jira-adapter/src/filters/workflow/types.ts [hunk 3]
```diff
@@ -167,15 +169,15 @@ export type Workflow = {
...
+  transitions: Transition[]
   statuses?: Status[]
 }
 
 export const workflowSchema = Joi.object({
   id: idSchema.optional(),
   entityId: Joi.string().optional(),
   name: Joi.string().optional(),
-  transitions: Joi.array().items(transitionsSchema).optional(),
+  transitions: Joi.array().items(transitionsSchema),
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/workflow/types.ts:180]: I think the default is optional. You need to put here `.required()`

File: packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts [hunk 4]
```diff
@@ -114,7 +116,7 @@ export const deployWorkflow = async (
       await deployTriggers(resolvedChange, client)
       // No need to run in DC since the main deployment requests already supports deploying steps
       if (!client.isDataCenter) {
-        await deploySteps(getChangeData(change), client)
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/workflow/workflow_deploy_filter.ts:117]: note that `getChangeData(change)` is the unresolved instance while `instance` is resolved. I think `deploySteps` expects the unresolved instance

File: packages/jira-adapter/test/filters/workflow/workflow_deploy_filter.test.ts [hunk 5]
```diff
@@ -181,7 +181,7 @@ describe('workflowDeployFilter', () => {
           .apiDefinitions.types.Workflow.deployRequests,
         fieldsToIgnore: expect.toBeFunction(),
       })
-    })
+    }, 100000)
```
Comment (alonstern) [inline packages/jira-adapter/test/filters/workflow/workflow_deploy_filter.test.ts]: Is this on purpose?

**Review by alonstern — APPROVED** [review]
🔥 Just check the comment before merging

**Pull request merged**